Juan Miguel Armenta Cano

In [43]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.axes as ax
import seaborn as sns
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import BaggingClassifier
from mlxtend.classifier import StackingCVClassifier
from sklearn.ensemble import VotingClassifier
from sklearn import preprocessing
from sklearn.model_selection import StratifiedShuffleSplit
from sklearn.metrics import f1_score
from numpy import mean
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import RandomForestClassifier

In [2]:
#Cargamos el DataSet con los datos de las zarigüeyas 
data = pd.read_csv('possum.csv')
#Observamos los encabezados del DataFrame
data.head()

case  site  Pop  sex  age  hdlngth  skullw  totlngth  taill  footlgth  \
0     1     1  Vic    1  8.0     94.1    60.4      89.0   36.0      74.5   
1     2     1  Vic    0  6.0     92.5    57.6      91.5   36.5      72.5   
2     3     1  Vic    0  6.0     94.0    60.0      95.5   39.0      75.4   
3     4     1  Vic    0  6.0     93.2    57.1      92.0   38.0      76.1   
4     5     1  Vic    0  2.0     91.5    56.3      85.5   36.0      71.0   

   earconch   eye  chest  belly  
0      54.5  15.2   28.0   36.0  
1      51.2  16.0   28.5   33.0  
2      51.9  15.5   30.0   34.0  
3      52.2  15.2   28.0   34.0  
4      53.2  15.1   28.5   33.0

In [3]:
data = data.dropna()
#Verificamos nuevamente la información del DataFrame
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 101 entries, 0 to 103
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   case      101 non-null    int64  
 1   site      101 non-null    int64  
 2   Pop       101 non-null    object 
 3   sex       101 non-null    int64  
 4   age       101 non-null    float64
 5   hdlngth   101 non-null    float64
 6   skullw    101 non-null    float64
 7   totlngth  101 non-null    float64
 8   taill     101 non-null    float64
 9   footlgth  101 non-null    float64
 10  earconch  101 non-null    float64
 11  eye       101 non-null    float64
 12  chest     101 non-null    float64
 13  belly     101 non-null    float64
dtypes: float64(10), int64(3), object(1)
memory usage: 11.8+ KB


In [4]:
#Agrupamos el conjunto de entrenamiento y variable de clase
X_train = data[['case', 'site', 'sex','age','hdlngth','skullw','totlngth','taill', 'footlgth' , 'earconch' , 'eye' , 'chest' , 'belly']]
y_train = data[['sex']]

In [5]:
from sklearn.datasets import make_circles, make_classification, make_moons
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
from sklearn.ensemble import AdaBoostClassifier, RandomForestClassifier
from sklearn.gaussian_process import GaussianProcessClassifier
from sklearn.gaussian_process.kernels import RBF
from sklearn.inspection import DecisionBoundaryDisplay
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_predict
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import BaggingClassifier
from mlxtend.classifier import StackingCVClassifier
from sklearn.ensemble import VotingClassifier
from sklearn.metrics import (confusion_matrix, accuracy_score)
import warnings
from sklearn.feature_selection import mutual_info_classif
from sklearn.decomposition import PCA
warnings.simplefilter("ignore")

Primero objetivo

se usaran 10 clasificadores los cuales son:
1. Arboles de desicion 
2. Aprendizaje bayesioano
3. Máquina de soporte vectorial
4. SVC
5. GaussianNB
6. RandomForest
7. AdaBoost
8. RBF
9. GaussianProcess
10. QDA

en todos estos clasificadores se usara validacion cruzada con K=10 usando cross_val_score


In [6]:
clfd1 = DecisionTreeClassifier(max_depth=5, random_state=42)
prediccion = cross_val_predict(clfd1, X_train, y_train, cv=10)
# confusion matrix
cm = confusion_matrix(y_train, prediccion) 
print ("Confusion Matrix : \n", cm) 
  
# accuracy score of the model
print('Test accuracy = ', accuracy_score(y_train, prediccion))
TP = cm[0][0]
FP = cm[0][1]
FN = cm[1][0]
TN = cm[1][1]
sensibilidad = TP / (TP + FN)

especificidad = TN / (TN + FP)
print('sensibilidad=' , sensibilidad)
print( 'especificidad=', especificidad)

Confusion Matrix : 
 [[42  0]
 [ 0 59]]
Test accuracy =  1.0
sensibilidad= 1.0
especificidad= 1.0


In [7]:
clfd2 = GaussianNB()
prediccion = cross_val_predict(clfd2, X_train, y_train, cv=10)
# confusion matrix
cm = confusion_matrix(y_train, prediccion) 
print ("Confusion Matrix : \n", cm) 
  
# accuracy score of the model
print('Test accuracy = ', accuracy_score(y_train, prediccion))
TP = cm[0][0]
FP = cm[0][1]
FN = cm[1][0]
TN = cm[1][1]
sensibilidad = TP / (TP + FN)

especificidad = TN / (TN + FP)
print('sensibilidad=' , sensibilidad)
print( 'especificidad=', especificidad)

Confusion Matrix : 
 [[42  0]
 [ 0 59]]
Test accuracy =  1.0
sensibilidad= 1.0
especificidad= 1.0


In [8]:
clfd3 = KNeighborsClassifier(5)
prediccion = cross_val_predict(clfd3, X_train, y_train, cv=10)
# confusion matrix
cm = confusion_matrix(y_train, prediccion) 
print ("Confusion Matrix : \n", cm) 
  
# accuracy score of the model
print('Test accuracy = ', accuracy_score(y_train, prediccion))
TP = cm[0][0]
FP = cm[0][1]
FN = cm[1][0]
TN = cm[1][1]
sensibilidad = TP / (TP + FN)

especificidad = TN / (TN + FP)
print('sensibilidad=' , sensibilidad)
print( 'especificidad=', especificidad)

Confusion Matrix : 
 [[11 31]
 [36 23]]
Test accuracy =  0.33663366336633666
sensibilidad= 0.23404255319148937
especificidad= 0.42592592592592593


In [9]:
clfd4 = SVC (kernel="linear", C=0.025, random_state=42)
prediccion = cross_val_predict(clfd3, X_train, y_train, cv=10)
# confusion matrix
cm = confusion_matrix(y_train, prediccion) 
print ("Confusion Matrix : \n", cm) 
  
# accuracy score of the model
print('Test accuracy = ', accuracy_score(y_train, prediccion))
TP = cm[0][0]
FP = cm[0][1]
FN = cm[1][0]
TN = cm[1][1]
sensibilidad = TP / (TP + FN)

especificidad = TN / (TN + FP)
print('sensibilidad=' , sensibilidad)
print( 'especificidad=', especificidad)

Confusion Matrix : 
 [[11 31]
 [36 23]]
Test accuracy =  0.33663366336633666
sensibilidad= 0.23404255319148937
especificidad= 0.42592592592592593


In [10]:
clfd5 = GaussianProcessClassifier (1.0 * RBF(1.0), random_state=42)
prediccion = cross_val_predict(clfd5, X_train, y_train, cv=10)
# confusion matrix
cm = confusion_matrix(y_train, prediccion) 
print ("Confusion Matrix : \n", cm) 
  
# accuracy score of the model
print('Test accuracy = ', accuracy_score(y_train, prediccion))
TP = cm[0][0]
FP = cm[0][1]
FN = cm[1][0]
TN = cm[1][1]
sensibilidad = TP / (TP + FN)

especificidad = TN / (TN + FP)
print('sensibilidad=' , sensibilidad)
print( 'especificidad=', especificidad)

Confusion Matrix : 
 [[28 14]
 [22 37]]
Test accuracy =  0.6435643564356436
sensibilidad= 0.56
especificidad= 0.7254901960784313


In [11]:
clfd6 =  RandomForestClassifier(max_depth=5, n_estimators=10, max_features=1, random_state=42)
prediccion = cross_val_predict(clfd6, X_train, y_train, cv=10)
# confusion matrix
cm = confusion_matrix(y_train, prediccion) 
print ("Confusion Matrix : \n", cm) 
  
# accuracy score of the model
print('Test accuracy = ', accuracy_score(y_train, prediccion))
TP = cm[0][0]
FP = cm[0][1]
FN = cm[1][0]
TN = cm[1][1]
sensibilidad = TP / (TP + FN)

especificidad = TN / (TN + FP)
print('sensibilidad=' , sensibilidad)
print( 'especificidad=', especificidad)

Confusion Matrix : 
 [[31 11]
 [ 3 56]]
Test accuracy =  0.8613861386138614
sensibilidad= 0.9117647058823529
especificidad= 0.835820895522388


In [12]:
clfd7 =  AdaBoostClassifier(random_state=42)
prediccion = cross_val_predict(clfd7, X_train, y_train, cv=10)
# confusion matrix
cm = confusion_matrix(y_train, prediccion) 
print ("Confusion Matrix : \n", cm) 
  
# accuracy score of the model
print('Test accuracy = ', accuracy_score(y_train, prediccion))
TP = cm[0][0]
FP = cm[0][1]
FN = cm[1][0]
TN = cm[1][1]
sensibilidad = TP / (TP + FN)

especificidad = TN / (TN + FP)
print('sensibilidad=' , sensibilidad)
print( 'especificidad=', especificidad)

Confusion Matrix : 
 [[42  0]
 [ 0 59]]
Test accuracy =  1.0
sensibilidad= 1.0
especificidad= 1.0


In [13]:
clfd8 =  SVC(gamma=2, C=1, random_state=42)
prediccion = cross_val_predict(clfd8, X_train, y_train, cv=10)
# confusion matrix
cm = confusion_matrix(y_train, prediccion) 
print ("Confusion Matrix : \n", cm) 
  
# accuracy score of the model
print('Test accuracy = ', accuracy_score(y_train, prediccion))
TP = cm[0][0]
FP = cm[0][1]
FN = cm[1][0]
TN = cm[1][1]
sensibilidad = TP / (TP + FN)

especificidad = TN / (TN + FP)
print('sensibilidad=' , sensibilidad)
print( 'especificidad=', especificidad)

Confusion Matrix : 
 [[ 0 42]
 [ 0 59]]
Test accuracy =  0.5841584158415841
sensibilidad= nan
especificidad= 0.5841584158415841


In [14]:
clfd9 =  GaussianProcessClassifier(1.0 * RBF(1.0), random_state=42)
prediccion = cross_val_predict(clfd9, X_train, y_train, cv=10)
# confusion matrix
cm = confusion_matrix(y_train, prediccion) 
print ("Confusion Matrix : \n", cm) 
  
# accuracy score of the model
print('Test accuracy = ', accuracy_score(y_train, prediccion))
TP = cm[0][0]
FP = cm[0][1]
FN = cm[1][0]
TN = cm[1][1]
sensibilidad = TP / (TP + FN)

especificidad = TN / (TN + FP)
print('sensibilidad=' , sensibilidad)
print( 'especificidad=', especificidad)

Confusion Matrix : 
 [[28 14]
 [22 37]]
Test accuracy =  0.6435643564356436
sensibilidad= 0.56
especificidad= 0.7254901960784313


In [15]:
clfd10 =  QuadraticDiscriminantAnalysis()
prediccion = cross_val_predict(clfd10, X_train, y_train, cv=10)
# confusion matrix
cm = confusion_matrix(y_train, prediccion) 
print ("Confusion Matrix : \n", cm) 
  
# accuracy score of the model
print('Test accuracy = ', accuracy_score(y_train, prediccion))
TP = cm[0][0]
FP = cm[0][1]
FN = cm[1][0]
TN = cm[1][1]
sensibilidad = TP / (TP + FN)

especificidad = TN / (TN + FP)
print('sensibilidad=' , sensibilidad)
print( 'especificidad=', especificidad)

Confusion Matrix : 
 [[42  0]
 [ 0 59]]
Test accuracy =  1.0
sensibilidad= 1.0
especificidad= 1.0


### Primero Objetivo.-

Basandonos en los resultados ateriores y discriminando los clasificsafores con los que obtuvimos valores de "1" vamos a tomar en cuenta solo los siguentes resultados:

1. RandomForestClassifier
    Test accuracy =  0.8613861386138614
2. GaussianProcessClassifier
    Test accuracy =  0.6435643564356436
3. SVC
    Test accuracy =  0.5841584158415841
    


In [16]:
data.head()

case  site  Pop  sex  age  hdlngth  skullw  totlngth  taill  footlgth  \
0     1     1  Vic    1  8.0     94.1    60.4      89.0   36.0      74.5   
1     2     1  Vic    0  6.0     92.5    57.6      91.5   36.5      72.5   
2     3     1  Vic    0  6.0     94.0    60.0      95.5   39.0      75.4   
3     4     1  Vic    0  6.0     93.2    57.1      92.0   38.0      76.1   
4     5     1  Vic    0  2.0     91.5    56.3      85.5   36.0      71.0   

   earconch   eye  chest  belly  
0      54.5  15.2   28.0   36.0  
1      51.2  16.0   28.5   33.0  
2      51.9  15.5   30.0   34.0  
3      52.2  15.2   28.0   34.0  
4      53.2  15.1   28.5   33.0

In [17]:
data=data.drop(['Pop'],axis=1)

In [18]:
data.head()

case  site  sex  age  hdlngth  skullw  totlngth  taill  footlgth  earconch  \
0     1     1    1  8.0     94.1    60.4      89.0   36.0      74.5      54.5   
1     2     1    0  6.0     92.5    57.6      91.5   36.5      72.5      51.2   
2     3     1    0  6.0     94.0    60.0      95.5   39.0      75.4      51.9   
3     4     1    0  6.0     93.2    57.1      92.0   38.0      76.1      52.2   
4     5     1    0  2.0     91.5    56.3      85.5   36.0      71.0      53.2   

    eye  chest  belly  
0  15.2   28.0   36.0  
1  16.0   28.5   33.0  
2  15.5   30.0   34.0  
3  15.2   28.0   34.0  
4  15.1   28.5   33.0

In [30]:
X = data[['case', 'site', 'sex','age','hdlngth','skullw','totlngth','taill', 'footlgth' , 'earconch' , 'eye' , 'chest' , 'belly']]
y = data[['sex']]
print (X.shape)
print (y.shape)
y = np.array(y).reshape(101)
print (y.shape)

(101, 13)
(101, 1)
(101,)


In [31]:
X = preprocessing.StandardScaler().fit(X).transform(X)

In [32]:
sss = StratifiedShuffleSplit(n_splits=5, test_size=0.2, random_state=0)

Acontinuacion se presenta el modelo base

In [33]:
clf_nb = GaussianNB()
scores = []

for train_index, test_index in sss.split(X, y):
    X_train, X_test = X[train_index], X[test_index]
    y_train, y_test = y[train_index], y[test_index]
    clf_nb.fit(X_train, y_train)
    y_pred = clf_nb.predict(X_test)
    scores.append(f1_score(y_test, y_pred, average='weighted'))

print(scores)
print(mean(scores))

[1.0, 1.0, 1.0, 1.0, 1.0]
1.0


### Boosting


In [40]:
clf = GradientBoostingClassifier(n_estimators=10)
scores = cross_val_score(clf, X, y, cv=10)
print(scores)
print(scores.mean())

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
1.0


### Bagging

clf = BaggingClassifier(n_estimators=10)
scores = cross_val_score(clf, X, y, cv=10)
print(scores)
print(scores.mean())

### Voting

In [44]:
clf1 = DecisionTreeClassifier()
clf2 = GaussianNB()
clf3 = KNeighborsClassifier(n_neighbors=3)
clf4 = svm.SVC()
eclf = VotingClassifier(
     estimators=[('dt', clf1), ('nb', clf2), ('kNN', clf3), ('svm', clf4)],
     voting='hard') # 'soft'
for clf, label in zip([clf1, clf2, clf3, clf4, eclf], ['dt', 'nb', 'kNN', 'svm', 'Ensemble']):
    scores = cross_val_score(clf, X, y, scoring='accuracy', cv=10)
    print("Accuracy: %0.2f (+/- %0.2f) [%s]" % (scores.mean(), scores.std(), label))

Accuracy: 1.00 (+/- 0.00) [dt]
Accuracy: 1.00 (+/- 0.00) [nb]
Accuracy: 0.94 (+/- 0.05) [kNN]
Accuracy: 1.00 (+/- 0.00) [svm]
Accuracy: 1.00 (+/- 0.00) [Ensemble]


## 2da. parte 

### Boosting

In [45]:
clf = GradientBoostingClassifier(n_estimators=10)
scores = cross_val_score(clf, X, y, cv=10)
print(scores)
print(scores.mean())

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
1.0


In [46]:
clf = GradientBoostingClassifier(n_estimators=50)
scores = cross_val_score(clf, X, y, cv=10)
print(scores)
print(scores.mean())

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
1.0


In [49]:
clf = GradientBoostingClassifier(n_estimators=5)
scores = cross_val_score(clf, X, y, cv=10)
print(scores)
print(scores.mean())

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
1.0


In [53]:
clf = GradientBoostingClassifier(n_estimators=1)
scores = cross_val_score(clf, X, y, cv=10)
print(scores)
print(scores.mean())

[0.54545455 0.6        0.6        0.6        0.6        0.6
 0.6        0.6        0.6        0.5       ]
0.5845454545454546


In [55]:
clf = GradientBoostingClassifier(n_estimators=1000)
scores = cross_val_score(clf, X, y, cv=10)
print(scores)
print(scores.mean())

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
1.0


### Bagging

In [58]:
clf = BaggingClassifier(n_estimators=50)
scores = cross_val_score(clf, X, y, cv=10)
print(scores)
print(scores.mean())

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
1.0


In [59]:
clf = BaggingClassifier(n_estimators=100)
scores = cross_val_score(clf, X, y, cv=10)
print(scores)
print(scores.mean())

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
1.0


In [61]:
clf = BaggingClassifier(n_estimators=1000)
scores = cross_val_score(clf, X, y, cv=10)
print(scores)
print(scores.mean())

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
1.0


In [62]:
clf = BaggingClassifier(n_estimators=5)
scores = cross_val_score(clf, X, y, cv=10)
print(scores)
print(scores.mean())

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
1.0


In [63]:
clf = BaggingClassifier(n_estimators=3)
scores = cross_val_score(clf, X, y, cv=10)
print(scores)
print(scores.mean())

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
1.0


In [64]:
clf = BaggingClassifier(n_estimators=1)
scores = cross_val_score(clf, X, y, cv=10)
print(scores)
print(scores.mean())

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
1.0


### Voting


In [66]:
clf1 = svm.SVC() 
clf2 = KNeighborsClassifier(n_neighbors=10)
clf3 = GaussianNB() 
clf4 = DecisionTreeClassifier()
eclf = VotingClassifier(
     estimators=[('dt', clf1), ('nb', clf2), ('kNN', clf3), ('svm', clf4)],
     voting='hard') # 'soft'
for clf, label in zip([clf1, clf2, clf3, clf4, eclf], ['dt', 'nb', 'kNN', 'svm', 'Ensemble']):
    scores = cross_val_score(clf, X, y, scoring='accuracy', cv=10)
    print("Accuracy: %0.2f (+/- %0.2f) [%s]" % (scores.mean(), scores.std(), label))

Accuracy: 1.00 (+/- 0.00) [dt]
Accuracy: 0.95 (+/- 0.05) [nb]
Accuracy: 1.00 (+/- 0.00) [kNN]
Accuracy: 1.00 (+/- 0.00) [svm]
Accuracy: 1.00 (+/- 0.00) [Ensemble]


In [67]:
clf1 = svm.SVC() 
clf2 = KNeighborsClassifier(n_neighbors=50)
clf3 = GaussianNB() 
clf4 = DecisionTreeClassifier()
eclf = VotingClassifier(
     estimators=[('dt', clf1), ('nb', clf2), ('kNN', clf3), ('svm', clf4)],
     voting='hard') # 'soft'
for clf, label in zip([clf1, clf2, clf3, clf4, eclf], ['dt', 'nb', 'kNN', 'svm', 'Ensemble']):
    scores = cross_val_score(clf, X, y, scoring='accuracy', cv=10)
    print("Accuracy: %0.2f (+/- %0.2f) [%s]" % (scores.mean(), scores.std(), label))

Accuracy: 1.00 (+/- 0.00) [dt]
Accuracy: 0.87 (+/- 0.17) [nb]
Accuracy: 1.00 (+/- 0.00) [kNN]
Accuracy: 1.00 (+/- 0.00) [svm]
Accuracy: 1.00 (+/- 0.00) [Ensemble]


In [69]:
clf1 = svm.SVC() 
clf2 = KNeighborsClassifier(n_neighbors=90)
clf3 = GaussianNB() 
clf4 = DecisionTreeClassifier()
eclf = VotingClassifier(
     estimators=[('dt', clf1), ('nb', clf2), ('kNN', clf3), ('svm', clf4)],
     voting='hard') # 'soft'
for clf, label in zip([clf1, clf2, clf3, clf4, eclf], ['dt', 'nb', 'kNN', 'svm', 'Ensemble']):
    scores = cross_val_score(clf, X, y, scoring='accuracy', cv=10)
    print("Accuracy: %0.2f (+/- %0.2f) [%s]" % (scores.mean(), scores.std(), label))

Accuracy: 1.00 (+/- 0.00) [dt]
Accuracy: 0.58 (+/- 0.03) [nb]
Accuracy: 1.00 (+/- 0.00) [kNN]
Accuracy: 1.00 (+/- 0.00) [svm]
Accuracy: 1.00 (+/- 0.00) [Ensemble]


In [70]:
clf1 = svm.SVC() 
clf2 = KNeighborsClassifier(n_neighbors=100)
clf3 = GaussianNB() 
clf4 = DecisionTreeClassifier()
eclf = VotingClassifier(
     estimators=[('dt', clf1), ('nb', clf2), ('kNN', clf3), ('svm', clf4)],
     voting='hard') # 'soft'
for clf, label in zip([clf1, clf2, clf3, clf4, eclf], ['dt', 'nb', 'kNN', 'svm', 'Ensemble']):
    scores = cross_val_score(clf, X, y, scoring='accuracy', cv=10)
    print("Accuracy: %0.2f (+/- %0.2f) [%s]" % (scores.mean(), scores.std(), label))

Accuracy: 1.00 (+/- 0.00) [dt]
Accuracy: nan (+/- nan) [nb]
Accuracy: 1.00 (+/- 0.00) [kNN]
Accuracy: 1.00 (+/- 0.00) [svm]
Accuracy: nan (+/- nan) [Ensemble]


### Conclusion 

Al momento de optimizar los valores observanoms que existe como en el estimators "1" de Boosting, de lo contrario Bagging no tuvo ninguna variante.
por ultimo al utilizar Voting ya que es un ensable de 4 Clasificadores se observa una pequeña variante al usar un valor de "90" ya que al uzar un valor de "100" nos arroja un valor "nan"